### PR TITLE
Add server handling and compliance test cases for mixed input messages.

### PR DESCRIPTION
**Pull request by robshakir**
```
 * (M) chk/chk.go
  - Minor docstring fixes.
 * (M) compliance/compliance.go
  - Add test cases that cover the combinations of more than one of
    SessionParameters, AFTOperation and ElectionID being specified
    together, which is illegal.
 * (M) fluent/fluent.go
   - Add the ability to inject an arbitrary pre-formed message to
     the client. This is preferred rather than having fluent helpers
     that build invalid messages.
 * (M) server/server.go
   - Ensure that the reference server handles the cases described
     above cleanly.
```

This was discovered in internal testing where it was found that the reference server never converges in some of these cases.
Specification clarifications are made in a gRIBI PR [here](https://github.com/openconfig/gribi/pull/86).

### CODE REVIEW

**Review by alshabib — APPROVED** [review]

File: fluent/fluent.go [hunk 1]
```diff
@@ -408,6 +408,15 @@ type gRIBIModify struct {
 	parent *GRIBIClient
 }
 
+// InjectRequest injects a gRIBI ModifyRequest that is created by an external
+// entity into the modify stream. No validation of the input message is performed.
+// It is intended to allow for invalid messages that the fluent library does not
+// allow the creation of to be sent to a server.
+func (g *gRIBIModify) InjectRequest(t testing.TB, m *spb.ModifyRequest) *gRIBIModify {
```
Comment (xw-g) [inline fluent/fluent.go:415]: reminds me again, why we want to return `*gRIBIModify`?

File: compliance/compliance.go [hunk 2]
```diff
@@ -537,6 +555,133 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
...
+					Index:   1,
+					NextHop: &aftpb.Afts_NextHop{},
+				},
+			},
+		}},
+	})
+
+	err := awaitTimeout(context.Background(), c, t, time.Minute)
+	if err == nil {
+		t.Fatalf("did not get expected error from server, got: nil")
```
Comment (xw-g) [inline compliance/compliance.go:594]: `t.Fatal`

File: compliance/compliance.go [hunk 3]
```diff
@@ -537,6 +555,133 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
...
+			High: 0,
+		},
+		Params: &spb.SessionParameters{
+			Redundancy: spb.SessionParameters_ALL_PRIMARY,
+		},
+	})
+
+	err := awaitTimeout(context.Background(), c, t, time.Minute)
+	if err == nil {
+		t.Fatalf("did not get expected error from server, got: nil")
```
Comment (xw-g) [inline compliance/compliance.go:629]: Same `t.Fatal`

File: compliance/compliance.go [hunk 4]
```diff
@@ -537,6 +555,133 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
...
+
+	// Inject a specfic invalid entry that specifies election ID in two places along
+	// with an entry.
+	c.Modify().InjectRequest(t, &spb.ModifyRequest{
+		ElectionId: &spb.Uint128{
+			Low:  electionID.Load(),
+			High: 0,
+		},
+		Params: &spb.SessionParameters{
+			Redundancy: spb.SessionParameters_ALL_PRIMARY,
```
Comment (xw-g) [inline compliance/compliance.go:623]: Let's use the most popular and implemented `SINGLE_PRIMARY` and `PRESERVE` mode for testing.

File: compliance/compliance.go [hunk 5]
```diff
@@ -537,6 +555,133 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
...
+	c.Start(context.Background(), t)
+	defer c.Stop(t)
+
+	c.StartSending(context.Background(), t)
+
+	// Inject a specfic invalid entry that specifies election ID in two places along
+	// with an entry.
+	c.Modify().InjectRequest(t, &spb.ModifyRequest{
+		Params: &spb.SessionParameters{
+			Redundancy: spb.SessionParameters_ALL_PRIMARY,
```
Comment (xw-g) [inline compliance/compliance.go:654]: Same as above.

File: compliance/compliance.go [hunk 6]
```diff
@@ -537,6 +555,133 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
...
+					Index:   1,
+					NextHop: &aftpb.Afts_NextHop{},
+				},
+			},
+		}},
+	})
+
+	err := awaitTimeout(context.Background(), c, t, time.Minute)
+	if err == nil {
+		t.Fatalf("did not get expected error from server, got: nil")
```
Comment (xw-g) [inline compliance/compliance.go:675]: Same as above.

File: fluent/fluent.go (hunk 1 above)
Comment (robshakir) [reply fluent/fluent.go:415]: This allows us to chain calls - so you can do `(*gRIBIModify).InjectRequest(...).AddEntry(...)`. In these cases it seems a little less likely that a caller would actually do this, but this is the fluent API pattern.

File: compliance/compliance.go [hunk 7]
```diff
@@ -143,6 +146,21 @@ var (
 			Fn:        ModifyConnectionSinglePrimaryPreserve,
 			ShortName: "Modify RPC Connection with invalid persist/redundancy parameters",
 		},
+	}, {
+		In: Test{
+			Fn:        InvalidElectionIDInTwoContexts,
```
[the remark this answers is not in the document]
Comment (robshakir) [reply compliance/compliance.go]: Makes sense to me, fixed.

File: compliance/compliance.go [hunk 8]
```diff
@@ -537,6 +555,133 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
 	chk.HasRecvClientErrorWithStatus(t, err, want, chk.AllowUnimplemented())
 }
 
+// InvalidElectionIDInTwoContexts ensures that the server returns an error when the client
+// attempts to update the election ID whilst simultaenously specifying an operation.
+func InvalidElectionIDInTwoContexts(c *fluent.GRIBIClient, t testing.TB, _ ...TestOpt) {
+	defer electionID.Add(2)
```
[the remark this answers is not in the document]
Comment (robshakir) [reply compliance/compliance.go]: Ah, originally how I had done this then it occupied two election IDs, now it's just using one, fixed. Thanks.

File: compliance/compliance.go (hunk 2 above)
Comment (robshakir) [reply compliance/compliance.go]: Fixed, thanks.

File: compliance/compliance.go (hunk 4 above)
Comment (robshakir) [reply compliance/compliance.go:623]: Done, thanks.

File: compliance/compliance.go (hunk 3 above)
Comment (robshakir) [reply compliance/compliance.go]: Done.

File: compliance/compliance.go (hunk 5 above)
Comment (robshakir) [reply compliance/compliance.go]: Done.

File: compliance/compliance.go (hunk 6 above)
Comment (robshakir) [reply compliance/compliance.go]: Done.

**Review by robshakir — COMMENTED** [review]
Thanks for the review, addressed the comments. PTAL.

File: compliance/compliance.go [hunk 9]
```diff
@@ -537,6 +555,135 @@ func ModifyConnectionSinglePrimaryPreserve(c *fluent.GRIBIClient, t testing.TB,
...
+
+	// Inject a specfic invalid entry that specifies election ID in two places along
+	// with the session parameters.
+	c.Modify().InjectRequest(t, &spb.ModifyRequest{
+		ElectionId: &spb.Uint128{
+			Low:  electionID.Load(),
+			High: 0,
+		},
+		Params: &spb.SessionParameters{
+			Redundancy:  spb.SessionParameters_ALL_PRIMARY,
```
Comment (xw-g) [inline compliance/compliance.go:623]: You forgot to change it to `SINGLE_PRIMARY` :) (the most implemented mode)

**Review by xw-g — APPROVED** [review]

File: compliance/compliance.go (hunk 4 above)
Comment (xw-g) [reply compliance/compliance.go]: You forgot to change it to SINGLE_PRIMARY :) (the most implemented mode)

File: compliance/compliance.go (hunk 9 above)
Comment (robshakir) [reply compliance/compliance.go:623]: Ugh, sorry -- fixing now...

**Pull request merged**